src/sources/kubernetes_logs/mod.rs

```diff
@@ -0,0 +1,317 @@
+                file_server.run(file_source_tx, futures::compat::Compat01As03::new(shutdown));
+            result.expect("file server exited with an error")
+        });
+
+        let mut parser = parser::build();
+        let mut partial_events_merger = partial_events_merger::build(auto_partial_merge);
+
+        let events = file_source_rx.map(|(bytes, file)| {
+            emit!(KubernetesLogsEventReceived {
+                file: &file,
+                byte_size: bytes.len(),
+            });
+            create_event(bytes, file)
+        });
+        let events = events
+            .filter_map(move |event| futures::future::ready(parser.transform(event)))
+            .filter_map(move |event| futures::future::ready(partial_events_merger.transform(event)))
+            .map(move |mut event| {
+                if annotator.annotate(&mut event).is_none() {
+                    emit!(KubernetesLogsEventAnnotationFailed { event: &event });
+                }
+                event
+            });
+
+        let event_processing_loop = events.map(Ok).forward(out);
+
+        let (reflector_process_shutdown_signal_tx, reflector_process_shutdown_signal_rx) =
+            futures::channel::oneshot::channel();
+        pin_mut!(reflector_process);
+        let reflector_process_with_shutdown =
+            cancel_on_signal(reflector_process, reflector_process_shutdown_signal_rx);
+
+        use std::future::Future;
+        use std::pin::Pin;
+        let list: Vec<Pin<Box<dyn Future<Output = ()> + Send>>> = vec![
+            Box::pin(reflector_process_with_shutdown.map(|result| match result {
+                Ok(()) => info!(message = "reflector process completed gracefully"),
+                Err(error) => error!(message = "reflector process exited with an error", ?error),
+            })),
+            Box::pin(file_server_join_handle.map(|result| match result {
+                Ok(FileServerShutdown) => {
+                    info!(message = "file server completed gracefully");
+                    reflector_process_shutdown_signal_tx
+                        .send(())
+                        .expect("unable to trigger reflector shutdown");
+                }
+                Err(error) => error!(message = "file server exited with an error", ?error),
+            })),
+            Box::pin(event_processing_loop.map(|result| {
+                match result {
+                    Ok(()) => info!(message = "event processing loop completed gracefully"),
+                    Err(error) => error!(
+                        message = "event processing loop exited with an error",
+                        ?error
+                    ),
+                };
+            })),
+        ];
+        let mut futs: futures::stream::FuturesUnordered<_> = list.into_iter().collect();
+
+        while let Some(()) = futs.next().await {
+            trace!(message = "another future complete");
+        }
+
+        info!(message = "done");
+        Ok(())
```

Comment:
We should hold onto `ShutdownSignalToken` until this point so it's dropped when we go out of scope here because at this point the local shutdown procedure has finished.

It could be returned from `file_server.run` and then propagated as `Output = Option<ShutdownSignalToken>` in the upper `list` and then hold it when we get it in `while let Some(..) = futs.next().await`.

Reply:
The shutdown logic here is problematic, and I'll have to revisit it altogether.
One problem that was discovered is that if an error happens in one of the tasks that's not the file server, with the current design we don't have a way to communicate that to the file server.

I'll rewrite the code in such a way that we don't even pass the `shutdown` directly, but pass an internal termination request there. Then we'll be able to turn file server off either because of `shutdown` or other task error. And also drop the `shutdown` at the very end appropriately.

Reply:
I added proper lifecycle management, should be fixed now.